### PR TITLE
React 19

### PR DESCRIPTION
**Pull request by mattgperry**
This PR upgrades React to v19.

Fixes https://github.com/framer/motion/issues/2668

Current release: `12.0.0-alpha.1`

Because Framer will be on React 18 perhaps for a while, the current plan is to keep this PR open, release as `12.0.0`, keep `main` for `11.x` releases and rebase them into `12.x` for the time being.

## TODO

- [x] Upgrade React
- [x] Build success
- [x] Fix types
- [x] Fix SSR tests
- [x] Fix unit tests
- [x] Fix e2e tests
- [x] Test backwards compatibility with 18.x (tested - because of changes to types it doesn't look like it)
- [x] Merge in https://github.com/framer/motion/pull/2699 and restore 3d tests to CI
- [ ] Upgrade to React 19 final on release
- [ ] Release `12.0.0`

### CODE REVIEW

**Comment by eps1lon** [comment]
For types, I'd recommend starting with [`npx types-react-codemod preset-19 ~/path-to-framer-motion-typescript-files`](https://github.com/eps1lon/types-react-codemod/#preset-19). https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69022 has the full changelog.

File: yarn.lock [hunk 1]
```diff
@@ -3703,6 +3712,15 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@types/react@npm:types-react@beta":
```
Comment (eps1lon) [inline yarn.lock:3715]: You should use `resolutions` to enforce a single version of `@types/react` and `@types/react-dom` as suggested in the [upgrade guide](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#installing)

Multiple versions of `@types/react` is not supported just like multiple versions of `react` are not supported.

**Comment by mattgperry** [comment]
@eps1lon I did try that but it didn't make any changes. Luckily I think I'm making headway with the types now.

**Comment by eps1lon** [comment]
> @eps1lon I did try that but it didn't make any changes.

I just tried it in a fresh clone and it made changes to 25 files:

```bash
$ npx types-react-codemod@latest preset-19 .
Processing 779 files... 
Spawning 10 workers...
Results: 
0 errors
753 unmodified
1 skipped
25 ok
Time elapsed: 6.403seconds 
```

**Comment by mattgperry** [comment]
> > @eps1lon I did try that but it didn't make any changes.
> 
> I just tried it in a fresh clone and it made changes to 25 files:
> 
> ```shell
> $ npx types-react-codemod@latest preset-19 .
> Processing 779 files... 
> Spawning 10 workers...
> Results: 
> 0 errors
> 753 unmodified
> 1 skipped
> 25 ok
> Time elapsed: 6.403seconds 
> ```

Ah interesting, I must have done it wrong. I’ll give it another shot, thanks.

**Comment by izakfilmalter** [comment]
So hype for this.

**Comment by steve-marmalade** [comment]
Thanks for pushing on this. I'm very happy to report that `framer-motion@12.0.0-alpha.0` is working well in my project running the latest NextJS canary and the latest react-19 RC.

**Comment by steve-marmalade** [comment]
Hi @mattgperry , just curious, does:
>  - [x] Merge in https://github.com/framer/motion/pull/2699 and restore 3d tests to CI


mean that you've already made significant progress on Framer Motion 3d compatibility with React 19?


I ask because I remember seeing the following warning [in the docs](https://www.framer.com/motion/three-introduction/#installation), so it was a pleasant surprise.
> Warning: Framer Motion 3D is only compatible with React 18, with no timeline for React 19 compatibility.

**Comment by mattgperry** [comment]
@steve-marmalade Yeah I think 3D should be working now but that will rely on the deps defined in the `package.json`.

**Comment by krachtstefan** [comment]
> Thanks for pushing on this. I'm very happy to report that `framer-motion@12.0.0-alpha.0` is working well in my project running the latest NextJS canary and the latest react-19 RC.

@steve-marmalade I'm also the latest NextJS canary (`15.0.0-rc.0`) but have some TypeScript problems. Did you change anything else after switching to `framer-motion@12.0.0-alpha.0`? The API did not change, right? My animation works but TypeScript complains with:


```
JSX element type 'motion.div' does not have any construct or call signatures.
```

and

```
'motion.div' cannot be used as a JSX component.
  Its type 'ForwardRefComponent<HTMLElementType, HTMLMotionProps<HTMLElementType>> | undefined' is not a valid JSX element type.
    Type 'undefined' is not assignable to type 'ElementType'.
```

I basically just use a very basic `<motion.div ...`

**Comment by Apezdr** [comment]
Getting a weird bug where the opacity isn't transitioning when I refresh the page. Using NextJS. Not sure how to solve for it because this issue was introduced in upgrading the packages below. I am using framer motion in my template for the route, also using the `/src` directory with a master layout.js that doesn't use any framer motion in it.

`"next": "^15.0.0-rc.0",`
`"react": "^19.0.0-rc-df5f2736-20240712",`
`"react-dom": "^19.0.0-rc-df5f2736-20240712",`
`"framer-motion": "^12.0.0-alpha.0",`

**Comment by bdrtsky** [comment]
Playing with v12 and React 19. Nothing breaking with my old code yet, but I can't figure out how to handle Optimistic UI updates. Anyone tried to animate items coming from useOptimistic? They keep rendering twice, because of how optimistic works.

**Comment by tyteen4a03** [comment]
Any chance we can get a alpha-1 backporting fixes since alpha-0 release?

**Comment by mattgperry** [comment]
@tyteen4a03 You can upgrade to `12.0.0-alpha.1` now, this contains all the latest releases!

**Comment by lucascurti** [comment]
> Getting a weird bug where the opacity isn't transitioning when I refresh the page (but fixes itself if I hard refresh shift-refresh). Using NextJS. Not sure how to solve for it because this issue was introduced in upgrading the packages below. I am using framer motion in my template for the route, also using the `/src` directory with a master layout.js that doesn't use any framer motion in it.

@Apezdr In case you want to give it a try, I've tested v12.0.0-alpha.1 and this is fixed for me

**Comment by jeffijoe** [comment]
Thanks @mattgperry, that seems to fix all the issues I've been having, except for one: `popLayout` doesn't seem to have any effect. Are there still any gotchas to watch out for? I remember reading that `opacity` could be a problem but I tried removing opacity styles and still see the same issue in that the space that the exiting element occupied is not made available until the animation is done.

**Comment by bdrtsky** [comment]
useAnimationControls is not working in Framer Motion 12?

**Comment by xzilja** [comment]
For anyone experiencing type issues when using components like `<motion.div>` or wrapping elements with `motion.create` while using new ref system, this seems to work as a temporary fix on alpha.1

You can add other dom elements as needed and should declare this in some sort of global.d.ts file

```ts
import type { ForwardRefComponent, HTMLMotionProps } from 'framer-motion'
import type { ElementType } from 'react'

declare module 'framer-motion' {
	export const motion: Omit<typeof import('framer-motion').motion, 'div'> & {
		div: ForwardRefComponent<ElementType, HTMLMotionProps<ElementType>>
	}
}
```

**Comment by imjeehoo** [comment]
Maybe you need to update Current release to `12.0.0-alpha.2`

File: packages/framer-motion/src/render/dom/scroll/index.ts [hunk 2]
```diff
@@ -86,6 +86,7 @@ function needsElementTracking(options?: ScrollOptions) {
 
 function scrollFunction(onScroll: OnScroll, options: ScrollOptions) {
     if (isOnScrollWithInfo(onScroll) || needsElementTracking(options)) {
+        console.log(isOnScrollWithInfo(onScroll), needsElementTracking(options))
```
Comment (trevorblades) [inline packages/framer-motion/src/render/dom/scroll/index.ts:89]: Just a reminder that this should be removed before release! Apologies if you were already aware of this :3

**Comment by faradaytrs** [comment]
Stil blocked by https://github.com/motiondivision/motion/issues/2925

**Comment by headironc** [comment]
React 19 released! 😀

**Comment by mattgperry** [comment]
Can someone who’s using 12.0.0 check out 11.13.3 and let me know if that works just as well?

**Comment by faradaytrs** [comment]
> Can someone who’s using 12.0.0 check out 11.13.3 and let me know if that works just as well?

Screenshot 2024-12-10 at 10 03 32

**Comment by thomashexagon** [comment]
> Can someone who’s using 12.0.0 check out 11.13.3 and let me know if that works just as well?

image

**Comment by mattgperry** [comment]
That's odd, `HTMLAttributesWithoutMotionProps` shouldn't exist any more, let me take a look

**Comment by mattgperry** [comment]
@faradaytrs @thomashexagon can you check `node_modues/framer-motion/package.json` and double check that you have `11.13.3` installed as those types don't exist in `dist/index.d.ts` here: https://www.npmjs.com/package/framer-motion?activeTab=code

**Comment by mattgperry** [comment]
Reproducing on 19 locally I don't get the `className` error but do get the `onClick` `any` error.

**Comment by faradaytrs** [comment]
> Reproducing on 19 locally I don't get the `className` error but do get the `onClick` `any` error.

so should i check or you figured it out?

**Comment by mattgperry** [comment]
@faradaytrs you I believe are running the correct version, now I’m curious about the className

**Pull request closed without merging**